src/proxy/dns.rs

```diff
@@ -0,0 +1,271 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{HashMap, HashSet};
+use std::net::SocketAddr;
+use std::sync::{Arc, RwLock};
+use std::time::Instant;
+
+use crate::config::ProxyMode;
+use crate::state::ProxyState;
+
+use drain::Watch;
+use itertools::Itertools;
+use tracing::{info, trace, warn};
+
+use super::Error;
+use crate::dns;
+use crate::proxy::ProxyInputs;
+use crate::state::workload::Workload;
+
+use tokio::task::JoinHandle;
+
+use trust_dns_proto::op::{Message, MessageType, Query};
+use trust_dns_proto::rr::{Name, RecordType};
+use trust_dns_proto::serialize::binary::BinDecodable;
+use trust_dns_server::authority::MessageRequest;
+use trust_dns_server::server::{Protocol, Request};
+
+struct TaskContext {
+    task: tokio::task::JoinHandle<()>,
+    // monotonic task start time
+    start: Instant,
+    finished: bool,
+    // TODO: honor dns cache ttl?
+}
+
+// PollingDns roughly maps to STRICT/LOGICAL DNS in envoy to allow ambient support for workload entry
+// hostnames that are resolved using async polling from the dataplane.
+// This can be important in niche cases, like geographic dns resolution, as documented in
+// https://github.com/istio/istio/blob/7bf52db38c91c89a4d56d6a94f402fd9ddaf6465/pilot/pkg/serviceregistry/kube/conversion.go#L151-L155
+pub(super) struct PollingDns {
+    pi: ProxyInputs,
+    drain: Watch,
+    // workload UID to task
+    tasks: HashMap<String, TaskContext>,
+}
+
+// TODO(kdorosh) DRY the copied code following here and generally clean up any latest placeholder test values
```

Comment:
Is set_id a test value?

Reply:
yes; this code was copied from test code and still needs to be make production-ready and live in the right place. first im getting the on-demand skeleton flushed out, ci passing, and then cleaning up

Reply:
addressed, this code was removed